tools/dumppdf.py

```diff
@@ -128,51 +129,186 @@ def dumpallobjs(out, doc, mode=None):
     out.write('</pdf>')
     return
 
-# dumpoutline
+
+def resolve_dest(dest, doc):
+    if isinstance(dest, str):
+        dest = resolve1(doc.get_dest(dest))
+    elif isinstance(dest, PSLiteral):
+        dest = resolve1(doc.get_dest(dest.name))
+    if isinstance(dest, dict):
+        dest = dest['D']
+    return dest
+
+
+def get_page_number(dest, doc, pages, a):
+    pageno = None
+    if dest:
+        dest = resolve_dest(dest, doc)
+        pageno = pages[dest[0].objid]
+    elif a:
+        action = a.resolve()
+        if isinstance(action, dict):
+            subtype = action.get('S')
+            if subtype and repr(
+                    subtype) == '/GoTo' and action.get('D'):
+                dest = resolve_dest(action['D'], doc)
+                pageno = pages[dest[0].objid]
+    return pageno
+
+
+def extract_outline_info(pages, outlines, doc, outfp, dest_info):
+    for (level, title, dest, a, se) in outlines:
+        pageno = None
+        if dest:
+            dest = resolve_dest(dest, doc)
+            pageno = pages[dest[0].objid]
+        elif a:
+            action = a.resolve()
+            if isinstance(action, dict):
+                subtype = action.get('S')
+                if subtype and repr(
+                        subtype) == '/GoTo' and action.get('D'):
+                    dest = resolve_dest(action['D'], doc)
+                    pageno = pages[dest[0].objid]
+        s = q(title)
+        outfp.write('<outline level="%r" title="%s">\n' %
+                    (level, q(s)))
+        if (dest_info):
+            get_dest_info(dest, outfp)
+        if pageno is not None:
+            outfp.write('<pageno>%r</pageno>\n' % pageno)
+        outfp.write('</outline>\n')
+
+
+def get_dest_info(dest, outfp):
```

Comment:
Would be nice to know what this does, what it returns and what the arguments are